### PR TITLE
Improvements to distillation unit model

### PR DESCRIPTION
**Pull request by agarciadiego**
## Fixes
Fixes warning with tray replacing pressure_ref and temperature_ref

## Summary/Motivation:
Resolves two requests from the issue #96 

## Changes proposed in this PR:
-Separates out make_phase_split  methods as utility methods and uses it in condenser and reboiler unit models
-Utilizes util propagate_state

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: idaes/models_extra/column_models/column_util.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
```
[the remark this answers is not in the document]
Comment (andrewlee94) [reply idaes/models_extra/column_models/column_util.py:2]: @agarciadiego I believe this is the right place - at most I would just change the name to `util.py`.

A related note however - could we get some unit tests for this function - just set up a set of dummy models and test all the possible options/paths.

File: idaes/models_extra/column_models/condenser.py [hunk 2]
```diff
@@ -243,9 +243,28 @@ def build(self):
 
         self._make_ports()
 
+        def rule_reflux(self):
+            return self.reflux_ratio * (1 + self.reflux_ratio) ** -1
+
+        self.expr1 = Expression(rule=rule_reflux)
```
Comment (andrewlee94) [inline idaes/models_extra/column_models/condenser.py:249]: I think this needs a better name. Also note that you can do `@self.Expression` in the same way as `Constraints`.

File: idaes/models_extra/column_models/tests/test_utilities.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+        )
+
+    # Call make_phase_split function
+    make_phase_split(m, m.port)
+
+    # Check for expected results
+    for i in m.time:
+        for p in m.config.property_package.phase_list:
+            for j in m.config.property_package.component_list:
+                assert m.port.flow_mol_phase[i, p, j].is_expression_type()
```
Comment (andrewlee94) [inline idaes/models_extra/column_models/tests/test_utilities.py:114]: This doesn't really check that it is the expression you expect however. That said, checking the form of an expression can be difficult as the only way I know to really do it is to check the `str(expr.body)`.

You basically have two options - hard code the expected string (which is fragile as things might change in Pyomo that affect this), or build the expected expression manually and then compare to what you got from the method (which can be a lot of work for complex expressions).

File: idaes/models_extra/column_models/tests/test_utilities.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+                assert value(m.port.flow_mol_phase_comp[i, p, j]) == 1e-8
+                assert (
+                    m.port.mass_frac_phase_comp[i, p, j]
+                    is m.properties_out[i].mass_frac_phase_comp[p, j]
+                )
+
+
+class DummyData:
+    """
+    Class containing necessary data to generate ccomponent list
```
Comment (andrewlee94) [inline idaes/models_extra/column_models/tests/test_utilities.py:315]: Typo: component

File: idaes/models_extra/column_models/tests/test_utilities.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+    def __init__(self):
+        self.config = self.config()
+
+    class config:
+        def __init__(self):
+            self.property_package = self.property_package()
+
+        class property_package:
+            def component_list(self, list):
```
Comment (andrewlee94) [inline idaes/models_extra/column_models/tests/test_utilities.py:326]: Where do `list` and `list1` come from here? Also, I would suggest slightly more descriptive names to help readers understand the tests when this gets used.

**Review by andrewlee94 — CHANGES_REQUESTED** [review]
A few comments in line. Biggest thing is to remove the 4.6' file that slipped in somehow.

File: idaes/models_extra/column_models/condenser.py [hunk 6]
```diff
@@ -243,26 +243,25 @@ def build(self):
 
         self._make_ports()
 
+        @self.Expression(doc="Reflux calculations")
         def rule_reflux(self):
```
Comment (andrewlee94) [inline idaes/models_extra/column_models/condenser.py:247]: You should give this a more indicative name as whatever you use here will be the name of the resulting `Expression` object.

File: idaes/models_extra/column_models/util.py [hunk 7]
```diff
@@ -286,13 +286,7 @@ def rule_flow(model, t, p, i):
 
                 model.add_component("e_" + local_name + port.local_name, expr)
                 port.add(expr, k)
-            else:
```
Comment (andrewlee94) [inline idaes/models_extra/column_models/util.py]: Just to check, are you sure this is no longer needed?

File: idaes/models_extra/column_models/util.py [hunk 8]
```diff
@@ -336,9 +330,3 @@ def rule_enth(model, t):
                 ref = Reference(var)
                 setattr(model, "_" + k + "_" + port.local_name + "_ref", ref)
                 port.add(ref, k)
-            else:
```
Comment (andrewlee94) [inline idaes/models_extra/column_models/util.py]: As above.

**Review by andrewlee94 — COMMENTED** [review]
A couple of minor comments.

**Comment by andrewlee94** [comment]
Also, the `xfailed` tests still need to be addressed in `test_tray_column.py` (and maybe elsewhere).

File: idaes/models_extra/column_models/util.py (hunk 7 above)
Comment (agarciadiego) [reply idaes/models_extra/column_models/util.py]: @andrewlee94 maybe you could give me more insight but my understanfing is that `if "phase" not in local_name:` and `elif "phase" in local_name:` are the only options available as it is in the local_name or is not but I may be wrong

Comment (andrewlee94) [reply idaes/models_extra/column_models/util.py]: Those are the options, however in that case the `elif` really should just be an `else`. I was mostly asking as it is difficult to follow the logic in a big statement like this so I wanted to be sure there was no chance of an overrun error.

File: idaes/models_extra/column_models/tests/test_tray_column.py [hunk 9]
```diff
@@ -139,7 +135,7 @@ def btx_fctp(self):
         # Inlet feed conditions
         m.fs.unit.feed.flow_mol_comp[0, "benzene"].fix(20)
         m.fs.unit.feed.flow_mol_comp[0, "toluene"].fix(20)
-        m.fs.unit.feed.temperature.fix(368)
+        m.fs.unit.feed.temperature.fix(368.001)
```
Comment (andrewlee94) [inline idaes/models_extra/column_models/tests/test_tray_column.py:138]: I don't think this is a really good fix - it is just sidestepping the real issue.

Comment (agarciadiego) [reply idaes/models_extra/column_models/tests/test_tray_column.py:138]: I'll work on the scaling

**Review by andrewlee94 — APPROVED** [review]

File: idaes/models_extra/column_models/condenser.py [hunk 10]
```diff
@@ -243,9 +243,27 @@ def build(self):
...
+            return self.reflux_ratio * (1 + self.reflux_ratio) ** -1
+
         if self.config.condenser_type == CondenserType.totalCondenser:
 
-            self._make_splits_total_condenser()
+            make_phase_split(
+                self.control_volume,
+                port=self.reflux,
+                phase=self._liquid_set,
+                side_sf=(self.reflux_split_fraction),
```
Comment (jghouse88) [inline idaes/models_extra/column_models/condenser.py:256]: Minor: why is side_sf value in parenthesis?

File: idaes/models_extra/column_models/condenser.py [hunk 11]
```diff
@@ -243,9 +243,27 @@ def build(self):
...
+                equipmentType=CondenserType.totalCondenser,
+            )
+
+            make_phase_split(
+                self.control_volume,
+                port=self.distillate,
+                phase=self._liquid_set,
+                side_sf=(1 - self.reflux_split_fraction),
```
Comment (jghouse88) [inline idaes/models_extra/column_models/condenser.py:264]: Minor: same as previous comment regarding parenthesis.

File: idaes/models_extra/column_models/condenser.py [hunk 12]
```diff
@@ -243,9 +243,27 @@ def build(self):
...
+                equipmentType=CondenserType.totalCondenser,
+            )
 
             if self.config.temperature_spec == TemperatureSpec.atBubblePoint:
```
Comment (jghouse88) [inline idaes/models_extra/column_models/condenser.py:268]: @agarciadiego A note for the future. This needed to be changed to support the newer version of the square flash formulation where we removed the need for bubble and dew point. The change I made (and that is sitting on my computer) is to replace this instead with a sum{k_i*x_i = 1} constraint.

File: idaes/models_extra/column_models/condenser.py [hunk 13]
```diff
@@ -279,7 +297,29 @@ def rule_total_cond(self, t):
                 )
 
         else:
-            self._make_splits_partial_condenser()
+
+            make_phase_split(
+                self.control_volume,
+                port=self.reflux,
+                phase=self._liquid_set,
+                side_sf=(self.reflux_split_fraction),
```
Comment (jghouse88) [inline idaes/models_extra/column_models/condenser.py:305]: Minor: Same comment regarding parenthesis.

File: idaes/models_extra/column_models/reboiler.py [hunk 14]
```diff
@@ -269,9 +269,31 @@ def rule_boilup_ratio(self, t):
...
+        self.vapor_reboil = Port(
+            noruleinit=True,
+            doc="Vapor outlet stream that is returned to " "to the bottom tray.",
+        )
+
+        make_phase_split(
+            self.control_volume,
+            port=self.bottoms,
+            phase=self._liquid_set,
+            side_sf=1,
```
Comment (jghouse88) [inline idaes/models_extra/column_models/reboiler.py:286]: Why is 1 hard coded here? A comment here will be helpful.

File: idaes/models_extra/column_models/tray_column.py [hunk 15]
```diff
@@ -294,6 +299,26 @@ def build(self):
...
+                iscale.set_scaling_factor(var, 1e1)
+            if "_t1" in var.name:
+                iscale.set_scaling_factor(var, 1e-2)
+
+        for constr in self.component_data_objects(Constraint, descend_into=True):
+            if "pressure_drop_equation" in constr.name:
+                iscale.constraint_scaling_transform(constr, 1e-2, overwrite=False)
+            if "eq_P_vap" in constr.name:
+                iscale.constraint_scaling_transform(constr, 1e-2, overwrite=False)
+
```
Comment (jghouse88) [inline idaes/models_extra/column_models/tray_column.py:321]: Nice!

**Review by jghouse88 — APPROVED** [review]
Nicely done. LGTM. Had minor comments but more questions.

File: idaes/models_extra/column_models/tests/test_tray_column.py (hunk 9 above)
Comment (jghouse88) [reply idaes/models_extra/column_models/tests/test_tray_column.py]: Agree. Was this resolved?

File: idaes/models_extra/column_models/condenser.py (hunk 12 above)
Comment (agarciadiego) [reply idaes/models_extra/column_models/condenser.py:268]: added to issue #96

File: idaes/models_extra/column_models/condenser.py (hunk 10 above)
Comment (agarciadiego) [reply idaes/models_extra/column_models/condenser.py]: deleted parenthesis

File: idaes/models_extra/column_models/reboiler.py (hunk 14 above)
Comment (agarciadiego) [reply idaes/models_extra/column_models/reboiler.py:286]: Added a comment

File: idaes/models_extra/column_models/tests/test_tray_column.py (hunk 9 above)
Comment (agarciadiego) [reply idaes/models_extra/column_models/tests/test_tray_column.py]: Yes, fixed with scaling

**Pull request merged**